test/jdk/java/foreign/TestCleaner.java

```diff
@@ -0,0 +1,191 @@
+    public void testBadDoubleRegister(Supplier<Cleaner> cleanerFactory, SegmentFunction segmentFunction) {
+        MemorySegment segment = MemorySegment.allocateNative(10);
+        segment = segment.registerCleaner(cleanerFactory.get());
+        segment = segmentFunction.apply(segment);
+        try {
+            segment.registerCleaner(cleanerFactory.get()); // error here!
+            fail();
+        } catch (IllegalStateException ex) {
+            if (!segment.isAlive()) {
+                assertTrue(ex.getMessage().contains("This segment is already closed"));
+            } else {
+                assertTrue(ex.getMessage().contains("Already registered with a cleaner"));
+            }
+        }
+    }
+
+    enum SegmentFunction implements Function<MemorySegment, MemorySegment> {
+        IDENTITY(Function.identity()),
+        CLOSE(s -> { s.close(); return s; }),
+        SHARE(s -> { return s.share(); });
+
+        private final Function<MemorySegment, MemorySegment> segmentFunction;
+
+        SegmentFunction(Function<MemorySegment, MemorySegment> segmentFunction) {
+            this.segmentFunction = segmentFunction;
+        }
+
+        @Override
+        public MemorySegment apply(MemorySegment segment) {
+            return segmentFunction.apply(segment);
+        }
+    }
+
+    @DataProvider
+    static Object[][] segmentFunctions() {
+        Supplier<?>[] cleaners = {
+                (Supplier<Cleaner>)Cleaner::create,
+                (Supplier<Cleaner>)CleanerFactory::cleaner
+        };
+
+        SegmentFunction[] segmentFunctions = SegmentFunction.values();
+        Object[][] data = new Object[cleaners.length * segmentFunctions.length][3];
+
+        for (int cleaner = 0 ; cleaner < cleaners.length ; cleaner++) {
+            for (int segmentFunction = 0 ; segmentFunction < segmentFunctions.length ; segmentFunction++) {
+                data[cleaner + (cleaners.length * segmentFunction)] =
+                        new Object[] { cleaners[cleaner], segmentFunctions[segmentFunction] };
+            }
+        }
+
+        return data;
+    }
+
+    enum RegisterKind {
+        BEFORE,
+        AFTER;
+    }
+
+    @DataProvider
+    static Object[][] cleaners() {
+        Supplier<?>[] cleaners = {
+                (Supplier<Cleaner>)Cleaner::create,
+                (Supplier<Cleaner>)CleanerFactory::cleaner
+        };
+
+        RegisterKind[] kinds = RegisterKind.values();
+
+        SegmentFunction[] segmentFunctions = SegmentFunction.values();
+        Object[][] data = new Object[cleaners.length * kinds.length * segmentFunctions.length][3];
+
+        for (int kind = 0 ; kind < kinds.length ; kind++) {
+            for (int cleaner = 0 ; cleaner < cleaners.length ; cleaner++) {
+                for (int segmentFunction = 0 ; segmentFunction < segmentFunctions.length ; segmentFunction++) {
+                    data[kind + kinds.length * cleaner + (cleaners.length * kinds.length * segmentFunction)] =
```

Comment:
Using an `ArrayList<Object[]>` with `list.toArray(Object[][]::new)` would make this easier to read:
```java
List<Object[]> data = new ArrayList<>();
for (Registered kind : RegisterKind.values()) {
  for (Object cleaner : cleaners) {
    for (SegmentFunction segmentFunction : SegmentFunction.values()) {
      data.add(new Object[] {kind, cleaner, segmentFunction});
    }
  }
}
return data.toArray(Object[][]::new);
```